src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -58,57 +58,169 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
+            if (csharpSemanticRanges.HasValue)
+            {
+                csharpSemanticRanges.Value.Dispose();
+            }
+
+            return null;
+        }
+
+        using var combinedSemanticRanges = CombineSemanticRanges(razorSemanticRanges, csharpSemanticRanges is not null ? csharpSemanticRanges.Value.Object : null);
 
         // We return null when we have an incomplete view of the document.
         // Likely CSharp ahead of us in terms of document versions.
         // We return null (which to the LSP is a no-op) to prevent flashing of CSharp elements.
-        if (combinedSemanticRanges is null)
+        if (combinedSemanticRanges is null || !combinedSemanticRanges.HasValue)
         {
             _logger.LogWarning("Incomplete view of document. C# may be ahead of us in document versions.");
+
+            if (csharpSemanticRanges.HasValue)
+            {
+                csharpSemanticRanges.Value.Dispose();
+            }
+
             return null;
         }
 
-        var data = ConvertSemanticRangesToSemanticTokensData(combinedSemanticRanges, codeDocument);
+        var data = ConvertSemanticRangesToSemanticTokensData(combinedSemanticRanges.Value.Object, codeDocument);
         var tokens = new SemanticTokens { Data = data };
 
+        if (csharpSemanticRanges.HasValue)
+        {
+            csharpSemanticRanges.Value.Dispose();
+        }
+
         return tokens;
     }
 
-    private static List<SemanticRange>? CombineSemanticRanges(List<SemanticRange>? ranges1, List<SemanticRange>? ranges2)
+    private static PooledObject<List<SemanticRange>>? CombineSemanticRanges(List<SemanticRange>? razorRanges, List<SemanticRange>? csharpRanges)
     {
-        if (ranges1 is null || ranges2 is null)
+        if (razorRanges is null || csharpRanges is null)
         {
             // If we have an incomplete view of the situation we should return null so we avoid flashing.
             return null;
         }
 
-        var newList = new List<SemanticRange>(ranges1.Count + ranges2.Count);
-        newList.AddRange(ranges1);
-        newList.AddRange(ranges2);
-
         // Because SemanticToken data is generated relative to the previous token it must be in order.
-        // We have a guarantee of order within any given language server, but the interweaving of them can be quite complex.
-        // Rather than attempting to reason about transition zones we can simply order our ranges since we know there can be no overlapping range.
-        newList.Sort();
+        // We have a guarantee of order within any given language server, but when we translate from the ranges
+        // in the C# document to ranges into the Razor document we lose that guarantee.
+        // Having converted them to SemanticRange objects, we can simply do the final round of a merge sort.
+        var pooledList = ListPool<SemanticRange>.GetPooledObject(out var newList);
+        newList.Capacity = razorRanges.Count + csharpRanges.Count;
```

Comment:
Use `SetCapacityIfLarger`, or whatever the similarly-named-method is. If you happen to get something from the pool that is bigger, then you don't need to change capacity, so save an array creation.